crypto/rsa_extra/rsassa_pss_asn1_test.cc

```diff
@@ -0,0 +1,367 @@
+    0x86, 0x48, 0x86, 0xf7, 0x0d, 0x01, 0x01, 0x09, 0x30, 0x0b, 0x06, 0x09,
+    0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x04};
+
+static const int omit_salt_len = -1;
+
+struct PssParamsTestInput {
+  const uint8_t *der;
+  size_t der_len;
+  int expected_hash_nid;
+  int expected_mask_gen_nid;
+  int expected_mgf1_hash_nid;
+  int expected_salt_len;
+} kPssParamsTestInputs[] = {
+    {pss_sha1_salt_absent, sizeof(pss_sha1_salt_absent), NID_undef, NID_undef,
+     NID_undef, omit_salt_len},
+    {pss_sha224_salt_absent, sizeof(pss_sha224_salt_absent), NID_sha224,
+     NID_mgf1, NID_sha224, omit_salt_len},
+    {pss_sha256_salt_absent, sizeof(pss_sha256_salt_absent), NID_sha256,
+     NID_mgf1, NID_sha256, omit_salt_len},
+    {pss_sha384_salt_absent, sizeof(pss_sha384_salt_absent), NID_sha384,
+     NID_mgf1, NID_sha384, omit_salt_len},
+    {pss_sha512_salt_absent, sizeof(pss_sha512_salt_absent), NID_sha512,
+     NID_mgf1, NID_sha512, omit_salt_len},
+    {pss_sha256_salt_default, sizeof(pss_sha256_salt_default), NID_sha256,
+     NID_mgf1, NID_sha256, 20},
+    {pss_sha256_salt_30, sizeof(pss_sha256_salt_30), NID_sha256, NID_mgf1,
+     NID_sha256, 30},
+    {jdk_pss_sha256_salt_0, sizeof(jdk_pss_sha256_salt_0), NID_sha256, NID_mgf1,
+     NID_sha256, 0},
+    {jdk_pss_sha256_salt_absent, sizeof(jdk_pss_sha256_salt_absent), NID_sha256,
+     NID_mgf1, NID_sha256, omit_salt_len},
+    {jdk_pss_sha256_salt_30, sizeof(jdk_pss_sha256_salt_30), NID_sha256,
+     NID_mgf1, NID_sha256, 30},
+    {jdk_pss_sha224_salt_absent, sizeof(jdk_pss_sha224_salt_absent), NID_sha224,
+     NID_mgf1, NID_sha224, omit_salt_len},
+    {jdk_pss_sha384_salt_absent, sizeof(jdk_pss_sha384_salt_absent), NID_sha384,
+     NID_mgf1, NID_sha384, omit_salt_len},
+    {jdk_pss_sha512_salt_absent, sizeof(jdk_pss_sha512_salt_absent), NID_sha512,
+     NID_mgf1, NID_sha512, omit_salt_len},
+    {jdk_pss_sha256_mgf1_sha512, sizeof(jdk_pss_sha256_mgf1_sha512), NID_sha256,
+     NID_mgf1, NID_sha512, omit_salt_len},
+    {jdk_pss_sha1_mgf1_sha256, sizeof(jdk_pss_sha1_mgf1_sha256), NID_undef,
+     NID_mgf1, NID_sha256, omit_salt_len},
+    {jdk_pss_sha256_mgf1_sha1, sizeof(jdk_pss_sha256_mgf1_sha1), NID_sha256,
+     NID_undef, NID_undef, omit_salt_len},
+    {jdk_pss_sha1_mgf1_sha1_salt_30, sizeof(jdk_pss_sha1_mgf1_sha1_salt_30), NID_undef,
+     NID_undef, NID_undef, 30},
+};
+
+class RsassaPssTest: public testing::TestWithParam<PssParamsTestInput> {};
+
+TEST_P(RsassaPssTest, DecodeParamsDer) {
+  const auto &param = GetParam();
+  CBS params;
+  params.data = param.der;
+  params.len = param.der_len;
+  RSASSA_PSS_PARAMS *pss = nullptr;
+  ASSERT_TRUE(RSASSA_PSS_parse_params(&params, &pss));
+  // Holds ownership of heap-allocated RSASSA_PSS_PARAMS.
+  bssl::UniquePtr<RSASSA_PSS_PARAMS> pss_ptr(pss);
+  // Expect all bytes of params are used.
+  ASSERT_FALSE(CBS_len(&params));
+  // Validate Hash Algorithm of RSASSA-PSS-params.
+  if (param.expected_hash_nid) {
```

Comment:
Here you are expected `NID_undef` to always be 0.
Better to check that `param.foo != NID_undef`.
Same applies below in other if statements.

Reply:
Yes, I checked the `NID_undef` value. When it's `NID_undef`, it means the else block should be executed.

Reply:
I mean, above you are assuming that `NID_undef = 0` always.

I am just pointing out that if `NID_undef` ever changes value, then the check above will result in unexpected behaviour. Which is why it is better to be explicit about what value you expect `param.expected_hash_nid` to be.

Reply:
Will do.

Reply:
Addressed.